omnigibson/object_states/attached_to.py

```diff
@@ -0,0 +1,293 @@
+from omnigibson.object_states.update_state_mixin import UpdateStateMixin
+from omnigibson.object_states.object_state_base import BooleanState, RelativeObjectState
+from omnigibson.object_states.link_based_state_mixin import LinkBasedStateMixin
+from omnigibson.object_states.contact_bodies import ContactBodies
+from omnigibson.utils.constants import JointType
+from omnigibson.utils.usd_utils import create_joint
+from omnigibson.utils.ui_utils import create_module_logger, suppress_omni_log
+from omnigibson.utils.python_utils import classproperty
+
+# Create module logger
+log = create_module_logger(module_name=__name__)
+
+
+# Create settings for this module
+m = create_module_macros(module_path=__file__)
+
+m.ATTACHMENT_LINK_PREFIX = "attachment"
+
+m.DEFAULT_POSITION_THRESHOLD = 0.05  # 5cm
+m.DEFAULT_ORIENTATION_THRESHOLD = np.deg2rad(5.0)  # 5 degrees
+m.DEFAULT_JOINT_TYPE = JointType.JOINT_FIXED
+m.DEFAULT_BREAK_FORCE = 10000  # Newton
+m.DEFAULT_BREAK_TORQUE = 10000  # Newton-Meter
+
+
+class AttachedTo(RelativeObjectState, BooleanState, ContactSubscribedStateMixin, JointBreakSubscribedStateMixin, LinkBasedStateMixin):
+    """
+        Handles attachment between two rigid objects, by creating a fixed/spherical joint between self.obj (child) and
+        other (parent). At any given moment, an object can only be attached to at most one other object, i.e.
+        a parent can have multiple children, but a child can only have one parent. Note that only
+        child.states[AttachedTo].get_value(parent) will return True.
+
+        Subclasses ContactSubscribedStateMixin, JointBreakSubscribedStateMixin
+        on_contact function attempts to attach self.obj to other when a CONTACT_FOUND event happens
+        on_joint_break function breaks the current attachment
+    """
+
+    @classproperty
+    def metalink_prefix(cls):
+        """
+        Returns:
+            str: Unique keyword that defines the metalink associated with this object state
+        """
+        return m.ATTACHMENT_LINK_PREFIX
+
+    @staticmethod
+    def get_dependencies():
+        return RelativeObjectState.get_dependencies() + [ContactBodies]
+
+    def _initialize(self):
+        super()._initialize()
+        self.initialize_link_mixin()
+
+        # Reference to the parent object (DatasetObject)
+        self.parent = None
+
+        # Reference to the female meta link of the parent object (RigidPrim)
+        self.parent_link = None
+
+        # Mapping from the female meta link names of self.obj to their children (Dict[str, Optional[DatasetObject] = None])
+        self.children = {link_name: None for link_name in self.links if link_name.split("_")[1].endswith("F")}
+
+        # Cache of parent link candidates for other objects (Dict[DatasetObject, Dict[str, str]])
+        # Map from the male meta link names of self.obj to the correspounding female meta link names of other
```

Comment:
Isn't this `map from other object to male meta link names of self.obj to the correspounding female meta link names of other`?